### PR TITLE
Add is empty filter to search / api

### PR DESCRIPTION
**Pull request by colemanw**
Overview
-----------
This adds 2 new operators, "Is Empty" & "Not Empty" to SearchKit (and APIv4 in general). It removes the "Is Null / Not Null" operators from SearchKit because they are overly technical and have confusingly similar names to the new operators (they are still available in the API generally, just not the SearchKit UI).

Comments
---------------
This is a reviewer's cut of #19710

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
I think I hit this enotice too

api\v4\Entity\SavedSearchTest::testSmartGroupWithHaving
Undefined variable: field

/home/jenkins/bknix-dfl/build/core-19736-7rty9/web/sites/all/modules/civicrm/Civi/Api4/Query/Api4SelectQuery.php:450
/home/jenkins/bknix-dfl/build/core-19736-

**Comment by eileenmcnaughton** [comment]
Thanks @colemanw  - I see you have extended this to search kit - that is great but we do get the extra field shown when it should be hidden

image

@seamuslee001 @MegaphoneJon note that @colemanw has opted for the words 'Is Blank' rather than 'Is empty' per search builder and civireport - do you have any opinion on that?

**Comment by colemanw** [comment]
Ok I think I fixed the test. Your screenshot may be a caching thing on your local. It's working for me after clearing browser cache:
image

**Comment by colemanw** [comment]
@eileenmcnaughton I would be tempted to remove "Is Null" and "Not Null" from SearchKit. The distinction between empty and null would befuddle all but the most advanced SQL users. Even advanced techies would struggle with it, as SQL treats null values in a quirky way compared to most programming languages.

**Comment by colemanw** [comment]
Re: labels, the choices I can think of for EMPTY operators are

- **Is Empty** / **Not Empty**
- **Is Blank** / **Not Blank**
- **Is Nothing** / **Not Nothing**

Now that I think about it, I feel partial to the 3rd, because it's more general, and applies better to situations e.g. where you join on an activity and want to search for contacts _without_ that activity. "Activity ID - Is Nothing" isn't perfectly grammatical but reads better than the others.

**Comment by eileenmcnaughton** [comment]
I'm OK with the idea of removing IS NULL from the UI only and if there turns out to be demand for it we can put it back - I  think I'd like to see more opinions on this

**Comment by colemanw** [comment]
@eileenmcnaughton OK I've updated this with...

- Tests
- Support for array-based entities like Afform
- Support for Integer and Float field types (in which case `empty === 0`)
- Removed IS/NOT NULL operators from SearchKit.
- Set the label text back to "Empty" instead of "Blank" or "Nothing" because my alternatives aren't significantly better so let's just be consistent.

**Comment by wmortada** [comment]
This makes sense to me. Unless there is a strong case for keeping 'Is Null' this looks better from a user experience perspective.

**Comment by eileenmcnaughton** [comment]
This was in the dev digest so yes - merge-ready & merge in the middle of the week if no objections via the feedback

**Comment by eileenmcnaughton** [comment]
@colemanw I have been encouraging a few search-kit only things to go into 5.36 even thought it's RC - mostly because I'm trying to do some docs and want to get screen shots that will last a bit better

I think maybe mid this week I'd have to stop that but I think this is a candidate

File: Civi/Api4/Query/Api4SelectQuery.php [hunk 1]
```diff
@@ -468,11 +484,17 @@ protected function composeClause(array $clause, string $type) {
...
-    $sql_clause = \CRM_Core_DAO::createSQLFilter($fieldAlias, [$operator => $value]);
-    if ($sql_clause === NULL) {
-      throw new \API_Exception("Invalid value in $type clause for '$expr'");
+    if ($operator === 'IS EMPTY' || $operator === 'IS NOT EMPTY') {
+      // If field is not a string or number, this will pass through and use IS NULL/IS NOT NULL
+      $operator = str_replace('EMPTY', 'NULL', $operator);
+      if (in_array($field['data_type'] ?? NULL, ['String', 'Integer', 'Float'], TRUE)) {
+        $emptyVal = $field['data_type'] === 'String' ? '""' : '0';
+        $isEmptyClause = $operator === 'IS NULL' ? "= $emptyVal OR" : "<> $emptyVal AND";
+        return "$fieldAlias $isEmptyClause $fieldAlias $operator";
```
Comment (eileenmcnaughton) [inline Civi/Api4/Query/Api4SelectQuery.php:493]: I think you need to put () around this - although it didn't turn out to be the trick to making my search work

https://lab.civicrm.org/documentation/docs/user-en/-/merge_requests/467

Comment (eileenmcnaughton) [reply Civi/Api4/Query/Api4SelectQuery.php:493]: @colemanw if you add the () I'll merge this  - if we get more feedback we can tinker further with the labels but I think the bulk of the code is not going to change. I put a screenshot with empty into the docs

Comment (colemanw) [reply Civi/Api4/Query/Api4SelectQuery.php:493]: @eileenmcnaughton did you see any problematic sql caused by the lack of brackets? It appears to me like they're already getting added by the api4 query builder:
image

Comment (eileenmcnaughton) [reply Civi/Api4/Query/Api4SelectQuery.php:493]: @colemanw it was in the JOIN clauses I was not seeing brackets added - ie where you join on source IS EMPTY + some other critieria there are no parentheses

Comment (colemanw) [reply Civi/Api4/Query/Api4SelectQuery.php]: Well that's inconsistent. Ok I've added the brackets @eileenmcnaughton and maybe someday we'll clean that up.

**Pull request merged**